### PR TITLE
[infra] Restructure & initial work on requests

### LINKED ISSUE #56: Add Request Response Messaging
- [x] Create some type structure to route messages based on their type (Request vs. Information)
- [x] Support message handlers for both with auto-response from infra to Request messages
- [x] Add to the API

### PR DESCRIPTION
**Pull request by neeldug**
Fixes #56

### CODE REVIEW

**Comment by neeldug** [comment]
I think the easiest way to implement hard-typing would be to introduce some helpers to take messages, switch on their payload's concrete type and return the actual struct, using Go generics, but would love to have some input from others on this. @Razvan00Rusu @0x6770

**Comment by max-wickham** [comment]
I'm worried that agents will have to do a switch case on all messages now and check what the expected payload struct is for a given message type. Surely if agents are going to have to do this we should be doing it for them to enforce the message types being sent and received in response to requests and so that agents know the exact form of a message being received by them by its type.

Basically I think this method reduces the likely hood of bugs in infra maybe but increases the likelihood of problems in agent implementations.

**Comment by neeldug** [comment]
> I'm worried that agents will have to do a switch case on all messages now and check what the expected payload struct is for a given message type. Surely if agents are going to have to do this we should be doing it for them to enforce the message types being sent and received in response to requests and so that agents know the exact form of a message being received by them by its type.
> 
> Basically I think this method reduces the likely hood of bugs in infra maybe but increases the likelihood of problems in agent implementations.

I think this could be an issue — however, I was thinking we could avoid switch casing altogether by adding helpers as discussed above to get the exact struct that matches the enum and throw in the case of an error which would enable agent teams to work out errors. I think also the previous implementation would incur more bugs agent side as the API surface is enormous and allows illegal methods of accessing communication.

**Comment by max-wickham** [comment]
Yeah I agree that we should remove access to objects and methods that agents shouldn't be using, but If a helper gets the correct struct surely the agents will still have to perform a switch on the type of that struct? Also if we have to implements a helper for each message type why not just implement a method for each message type that calls the correct strategy method, that way the agents have less code to implement.

**Comment by max-wickham** [comment]
We can ensure the api for agents is smaller by ensuring they just have to imlplement the list of strategy methds and the only function they have access to is one to send messages to other agents.

**Comment by neeldug** [comment]
Not for each message type — and also having spent a couple hours on it yesterday, there was no solution to fixing the previous method since Communication would have to be exported from BaseAgent directly, it also became super confusing when going through refactors.

**Comment by neeldug** [comment]
> Yeah I agree that we should remove access to objects and methods that agents shouldn't be using, but If a helper gets the correct struct surely the agents will still have to perform a switch on the type of that struct? Also if we have to implements a helper for each message type why not just implement a method for each message type that calls the correct strategy method, that way the agents have less code to implement.

This creates the circular dependency which incurs the API surface increasing

**Comment by neeldug** [comment]
> We can ensure the api for agents is smaller by ensuring they just have to imlplement the list of strategy methds and the only function they have access to is one to send messages to other agents.

I agree, but doing it such that the messages require the agent strategy means that the circular dependency occurs.

File: pkg/infra/game/agent/agent.go [hunk 1]
```diff
@@ -13,8 +12,9 @@ import (
 )
 
 type Strategy interface {
-	HandleFightMessage(m message.TaggedMessage, view *state.View, agent BaseAgent, log *immutable.Map[commons.ID, decision.FightAction]) decision.FightAction
-	Default() decision.FightAction
+	HandleFightInformation(m message.TaggedMessage, view *state.View, agent BaseAgent, log *immutable.Map[commons.ID, decision.FightAction])
+	HandleFightRequest(m message.TaggedMessage, view *state.View, log *immutable.Map[commons.ID, decision.FightAction]) message.Payload
```
Comment (Razvan00Rusu) [inline pkg/infra/game/agent/agent.go:16]: Are we going to have 2 of these methods for every type of message that needs to be sent. This seems very verbose and isn't intuitive as to what message should be sent via which method. I understand the rationale for it but is there a way we could make it more explicit somehow?

**Review by Razvan00Rusu — APPROVED** [review]
Apart from my comment above (if we can think of a more elegant solution that would be great :D ) - LGTM

**Pull request merged**